File: backend/app/routers/events.py
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import require_admin
from app.models import Event
from app.schemas import EventCreate, EventOut, EventUpdate

router = APIRouter(tags=["events"])


@router.get("/api/events", response_model=list[EventOut])
def list_events(db: Session = Depends(get_db)):
    return db.query(Event).order_by(Event.event_datetime.asc()).all()


@router.get("/api/events/upcoming", response_model=list[EventOut])
def upcoming_events(limit: int = 3, db: Session = Depends(get_db)):
    return (
        db.query(Event)
        .filter(Event.event_datetime >= datetime.utcnow())
        .order_by(Event.event_datetime.asc())
        .limit(limit)
        .all()
    )


@router.post("/api/admin/events", response_model=EventOut, status_code=status.HTTP_201_CREATED)
def create_event(payload: EventCreate, db: Session = Depends(get_db), _admin=Depends(require_admin)):
    event = Event(**payload.model_dump())
    db.add(event)
    db.commit()
    db.refresh(event)
    return event


@router.put("/api/admin/events/{event_id}", response_model=EventOut)
def update_event(
    event_id: str,
    payload: EventUpdate,
    db: Session = Depends(get_db),
    _admin=Depends(require_admin),
):
    event = db.query(Event).filter(Event.id == event_id).first()
    if not event:
        raise HTTPException(status_code=404, detail="Event not found")
    for field, value in payload.model_dump().items():
        setattr(event, field, value)
    db.commit()
    db.refresh(event)
    return event


@router.delete("/api/admin/events/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_event(event_id: str, db: Session = Depends(get_db), _admin=Depends(require_admin)):
    event = db.query(Event).filter(Event.id == event_id).first()
    if not event:
        raise HTTPException(status_code=404, detail="Event not found")
    db.delete(event)
    db.commit()
    return None

File: frontend/src/services/events.ts
import { api } from './api'
import type { FestivalEvent } from '../types'

export async function getAllEvents(): Promise<FestivalEvent[]> {
  const { data } = await api.get<FestivalEvent[]>('/api/events')
  return data
}

export async function getUpcomingEvents(limit = 3): Promise<FestivalEvent[]> {
  const { data } = await api.get<FestivalEvent[]>('/api/events/upcoming', { params: { limit } })
  return data
}

export interface EventPayload {
  title: string
  description: string
  event_datetime: string
}

export async function createEvent(payload: EventPayload): Promise<FestivalEvent> {
  const { data } = await api.post<FestivalEvent>('/api/admin/events', payload)
  return data
}

export async function updateEvent(id: string, payload: EventPayload): Promise<FestivalEvent> {
  const { data } = await api.put<FestivalEvent>(`/api/admin/events/${id}`, payload)
  return data
}

export async function deleteEvent(id: string): Promise<void> {
  await api.delete(`/api/admin/events/${id}`)
}
